src/compiler/utilities.ts

```diff
@@ -1405,7 +1405,7 @@ namespace ts {
     }
 
     // True if the given identifier, string literal, or number literal is the name of a declaration node
-    export function isDeclarationName(name: Node): name is Identifier | StringLiteral | LiteralExpression {
+    export function isDeclarationName(name: Node): name is (Identifier & { __weakTypeGuard: void; }) | StringLiteral | LiteralExpression {

```

Comment:
If a type guard returns false, it is expected that the argument is not of the specified type. However, `isDeclarationName` can return false when `name` is an identifier. This wasn't an issue, as `isDeclarationName` was never used in an if statement with an else. However, with control flow based type guards, this wasn't working anymore. The then-block of an if statement contained a return, thus the statements after the if were the else branch of the if. It might not be the best way to solve this, but it's working..